8thwall-camera.js

```diff
@@ -5,88 +5,402 @@
+
+        XR8.XrController.updateCameraProjectionMatrix({
+            origin: { x: pos[0], y: pos[1], z: pos[2] },
+            facing: { x: rot[0], y: rot[1], z: rot[2], w: rot[3] },
+            // cam: { pixelRectWidth: Module.canvas.width, pixelRectHeight: Module.canvas.height, nearClipPlane: 0.01, farClipPlane: 100 }
+        })
+
+        WL.scene.onPreRender.push(() => {
+            gl.bindFramebuffer(gl.DRAW_FRAMEBUFFER, null); // <--- Should not be needed after next nightly is released (current 20230110)
+            XR8.runPreRender(Date.now());
+            XR8.runRender(); // <--- tell 8thwall to do it's thing (alternatively call this.GlTextureRenderer.onRender() if you only care about camera feed )
+        });
+
+        WL.scene.onPostRender.push(() => {
+            XR8.runPostRender(Date.now())
+        });
+    },
+
+    onCameraStatusChange: function (e) {
+        if (e && e.status === "failed") {
+            this.onException(new Error(`Camera failed with status: ${e.status}`));
         }
     },
 
-    update: function() {
-        if(this.started) {
-            if(WL.scene.onPostRender.length == 0) {
-                WL.scene.onPreRender.push(function() {
-                    XR8.runPreRender(Date.now());
-                    _wl_reset_context();
-                });
-                WL.scene.onPostRender.push(function() {
-                    XR8.runPostRender(Date.now());
-                    //_wl_reset_context();
-                });
-            }
+    /**
+     * @param {*} e 
+     * 
+     * private, called by 8thwall
+     */
+    onUpdate: function (e) {
+        if (!e.processCpuResult.reality)
+            return;
+
+        const { rotation, position, intrinsics } = e.processCpuResult.reality;
+
+        this.rotation[0] = rotation.x;
+        this.rotation[1] = rotation.y;
+        this.rotation[2] = rotation.z;
+        this.rotation[3] = rotation.w;
 
-            if(this.rotation[0] == 0 && this.rotation[1] == 0 &&
-                this.rotation[2] == 0 && this.rotation[3] == 0) {
-                return;
+        this.position[0] = position.x;
+        this.position[1] = position.y;
+        this.position[2] = position.z;
+
+        if (intrinsics) {
+            for (let i = 0; i < 16; i++) {
+                if (Number.isFinite(intrinsics[i])) { // some processCpuResult.reality.intrinsics are set to Infinity, which WL brakes our projectionMatrix. So we just filter those elements out
+                    this.view.projectionMatrix[i] = intrinsics[i];
+                }
             }
+        }
 
-            /* Apply the transform retrieved from 8thwall */
-            this.object.resetTransform();
-            this.object.rotate(this.rotation);
-            this.object.translate(this.position);
+        if (position && rotation) {
+            //this.object.resetTransform();
+            //this.object.rotate(this.rotation);
+            //this.object.translateWorld(this.position);
+
```

Comment:
You can probably remove this commented out code, it will remain in git history forever :)